infrastructures/infrastructure-factory/src/main/java/org/eclipse/che/api/factory/server/scm/kubernetes/KubernetesPersonalAccessTokenManager.java

```diff
@@ -144,19 +145,24 @@ public PersonalAccessToken fetchAndSave(Subject cheUser, String scmServerUrl)
           Map<String, String> annotations = secret.getMetadata().getAnnotations();
           if (annotations.get(ANNOTATION_CHE_USERID).equals(cheUser.getUserId())
               && annotations.get(ANNOTATION_SCM_URL).equals(scmServerUrl)) {
-            return Optional.of(
+            PersonalAccessToken token =
                 new PersonalAccessToken(
                     annotations.get(ANNOTATION_SCM_URL),
                     annotations.get(ANNOTATION_CHE_USERID),
                     annotations.get(ANNOTATION_SCM_USERNAME),
                     annotations.get(ANNOTATION_SCM_USERID),
-                    annotations.get(ANNOTATION_SCM_PERSONAL_ACCESS_TOKEN_ID),
                     annotations.get(ANNOTATION_SCM_PERSONAL_ACCESS_TOKEN_NAME),
-                    new String(Base64.getDecoder().decode(secret.getData().get("token")))));
+                    annotations.get(ANNOTATION_SCM_PERSONAL_ACCESS_TOKEN_ID),
+                    new String(Base64.getDecoder().decode(secret.getData().get("token"))));
+            if (scmPersonalAccessTokenFetcher.isValid(token)) {
+              return Optional.of(token);
+            } else {
+              clientFactory.create().secrets().inNamespace(namespaceMeta.getName()).delete(secret);
```

Comment:
It is not completely obvious to me why we should be deleting the invalid tokens here. Maybe a small comment for the reason behind this? Why don't we for example try `fetchAndSave` in this scenario instead?

Reply:
Explained in comment https://github.com/eclipse/che/pull/19518/commits/3c99d06df9531d65f9992c1ee01285a216432d0d